src/hooks/useDialog/useDialog.tsx

```diff
@@ -0,0 +1,111 @@
+import MessageDialog, { MessageDialogProps } from '@/components/Dialogs/MessageDialog'
+import { createId } from '@/utils/createId'
+import React, { useCallback, useEffect, useState, useContext, useMemo } from 'react'
+import { TransitionGroup } from 'react-transition-group'
+
+type DialogRendererProps = {
+  component: React.FunctionComponent
+}
+
+interface DialogsContainerProps {
+  dialogs: Record<string, React.FunctionComponent>
+  component?: React.ComponentType
+  container?: Element
+}
+
+const DialogsRenderer = ({ component, ...rest }: DialogRendererProps) => component(rest)
+
+export const DialogsContainer = ({ dialogs }: DialogsContainerProps) => {
+  return (
+    <TransitionGroup>
+      {Object.keys(dialogs).map((key) => (
+        <DialogsRenderer key={key} component={dialogs[key]} />
+      ))}
+    </TransitionGroup>
+  )
+}
+
+type DialogContextValue = {
+  openDialog: (key: string, dialog: React.FunctionComponent<{ in?: boolean }>) => void
+  closeDialog: (key: string) => void
+}
+const DialogContext = React.createContext<undefined | DialogContextValue>(undefined)
+DialogContext.displayName = 'DialogsContext'
+
+type DialogsProviderProps = {
+  container?: Element
+}
+
+export const DialogProvider: React.FC<DialogsProviderProps> = ({ container, children }) => {
+  const [dialogs, setDialogs] = useState<Record<string, React.FC>>({})
+
+  const openDialog = useCallback(
+    (key: string, dialog: React.FC) =>
+      setDialogs((dialogs) => ({
+        ...dialogs,
+        [key]: dialog,
+      })),
+    []
+  )
+
+  const closeDialog = useCallback(
+    (key: string) =>
+      setDialogs((dialogs) => {
+        if (!dialogs[key]) {
+          return dialogs
+        }
+        const newDialogs = { ...dialogs }
+        delete newDialogs[key]
+        return newDialogs
+      }),
+    []
+  )
+
+  const contextValue = useMemo(() => ({ openDialog, closeDialog }), [closeDialog, openDialog])
+
+  return (
+    <DialogContext.Provider value={contextValue}>
+      {children}
+      <DialogsContainer dialogs={dialogs} container={container} />
+    </DialogContext.Provider>
+  )
+}
+
+export const useDialog = (dialogProps?: MessageDialogProps) => {
+  const ctx = useContext(DialogContext)
+  if (ctx === undefined) {
+    throw new Error('useDialog must be used within a DialogProvider')
+  }
+
+  const { openDialog, closeDialog } = ctx
+
+  const [isShown, setShown] = useState(false)
+  const showModal = useCallback(() => setShown(true), [])
+  const hideModal = useCallback(() => setShown(false), [])
+  const dialogId = useMemo(() => createId(), [])
```

Comment:
_In theory_, we shouldn't use `useMemo` this way, docs state there is no guarantee that this will not re-run (it won't in current version of react but still). I think we could just use this instead?
```suggestion
  const dialogId = useRef(createId()).current
```